modality-probe-cli/src/export/mod.rs

```diff
@@ -0,0 +1,340 @@
+/// collected coumnar form as input.
+#[derive(Debug, PartialEq, StructOpt)]
+pub struct Export {
+    /// Generate the graph showing only the causal relationships,
+    /// eliding the events inbetween.
+    #[structopt(short, long)]
+    pub interactions_only: bool,
+    /// The path to a component directory. To include multiple
+    /// components, provide this switch multiple times.
+    #[structopt(short, long, required = true)]
+    pub components: Vec<PathBuf>,
+    /// The path to the collected trace.
+    #[structopt(short, long, required = true)]
+    pub report: PathBuf,
+    /// The type of graph to output.
+    ///
+    /// This can be either `cyclic` or `acyclic`.
+    ///
+    /// * A cyclic graph is one which shows the possible paths from
+    /// either an event or a probe.
+    ///
+    /// * An acyclic graph shows the causal history of either all
+    /// events or the interactions between traces in the system.
+    #[structopt(required = true)]
+    pub graph_type: GraphType,
+}
+
+#[derive(Debug, PartialEq, StructOpt)]
+pub enum GraphType {
+    Cyclic,
+    Acyclic,
+}
+
+impl FromStr for GraphType {
+    type Err = ExportError;
+    fn from_str(s: &str) -> Result<Self, ExportError> {
+        match s {
+            "cyclic" => Ok(GraphType::Cyclic),
+            "acyclic" => Ok(GraphType::Acyclic),
+            _ => Err(ExportError(format!("{} is not a valid graph type", s))),
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+#[error(display = "{}", _0)]
+pub struct ExportError(String);
+
+impl From<fmt::Error> for ExportError {
+    fn from(e: fmt::Error) -> Self {
+        ExportError(format!("encountered an error generating dot: {}", e))
+    }
+}
+
+impl From<CollectorError> for ExportError {
+    fn from(e: CollectorError) -> Self {
+        ExportError(format!("failed to read log file: {}", e))
+    }
+}
+
+pub fn run(mut exp: Export) -> Result<(), ExportError> {
+    let cfg = assemble_components(&mut exp.components)?;
+    let mut log_file = File::open(&exp.report).map_err(|e| {
+        ExportError(format!(
+            "failed to open the report file at {}: {}",
+            exp.report.display(),
+            e
+        ))
+    })?;
+
+    let graph = graph::log_to_graph(
+        json::read_log_entries(&mut log_file)?
+            .into_iter()
+            .peekable(),
+    )?;
+
+    let mut out = String::new();
+    writeln!(out, "Digraph G {{")?;
+
+    match (exp.graph_type, exp.interactions_only) {
+        (GraphType::Acyclic, false) => println!("{}", graph.graph.to_dot(&cfg)?),
+        (GraphType::Acyclic, true) => println!("{}", graph.graph.into_interactions().to_dot(&cfg)?),
+        (GraphType::Cyclic, false) => println!("{}", graph.graph.into_states().to_dot(&cfg)?),
+        (GraphType::Cyclic, true) => println!("{}", graph.graph.into_topology().to_dot(&cfg)?),
+    }
+
+    Ok(())
+}
+
+struct Cfg {
+    pub probes: HashMap<u32, ProbeMeta>,
+    pub events: HashMap<(Uuid, u32), EventMeta>,
```

Comment:
It seems like this representation might fail to capture events that arrived from a probe not associated with a pre-known component.

Reply:
That is true. In a very early pass (pre-report redux), I just used "`UNKNOWN_EVENT`" in the metadata, I can go back that direction here.

Reply:
An alternate approach is to use an enum for keying instead of a Uuid. Maybe like:

```rust
enum {
    Component(Uuid),
    Internal,
    UnknownAtProbe(ProbeId),
}
```